DOCSP-46694 - Extended JSON (#169)
mongodb/docs-pymongo: source/data-formats/extended-json.txt

You can read an Extended JSON string into a Python object by calling
the ``bson.json_util.loads()`` method. This method parses an Extended
JSON string and returns a Python list containing the data.

The following example shows how you can read an example Extended JSON string into a
list of dictionaries by using the ``loads()`` method:

.. io-code-block::

   .. input::
      :language: python

      from bson.json_util import loads

      ejson_str = '''[
         {"foo": [1, 2]},
         {"bar": {"hello": "world"}},
         {"code": {
            "$scope": {},
            "$code": "function x() { return 1; }"
         }},
         {"bin": {
            "$type": "80",
            "$binary": "AQIDBA=="
         }}
      ]'''

      doc = loads(ejson_str)
      print(doc)

   .. output::
      :language: none
      :visible: false

      [
         {'foo': [1, 2]},
         {'bar': {'hello': 'world'}},
         {'code': Code('function x() { return 1; }', {})},
         {'bin': Binary(b'\x01\x02\x03\x04', 128)}
      ]

Write Extended JSON
-------------------

You can write an Extended JSON string from a list of dictionaries
by calling the ``bson.json_util.dumps()`` method. 

> In this example, we output the Extended JSON in the Relaxed mode format.

Author: Good question. I based this page on [our Java page](https://www.mongodb.com/docs/drivers/java/sync/current/fundamentals/data-formats/document-data-format-extended-json/), which doesn't use monospace, and neither does the [Server page](https://www.mongodb.com/docs/manual/reference/mongodb-extended-json/) it links to.

> In this example, we output the Extended JSON in the Relaxed mode format.

Author: ```suggestion
The following example outputs an Extended JSON string in the Relaxed format:
```

> In this example, we output the Extended JSON in the Relaxed mode format.

Author: Our other docs have more info on the formats (see the Java page I linked earlier), but I thought it was redundant to the Server manual, so I just went with a link. 

One option is to the move the link higher in the page and make it more visible?

> In this example, we output the Extended JSON in the Relaxed mode format.

Reviewer: Makes sense. Yeah I think making the link more visible and closer to the first mention would be helpful. Maybe linking it inline on first mention (although I know we don't do that often) or just adding a "To learn more about..." somewhere closer

> In this example, we output the Extended JSON in the Relaxed mode format.

Reviewer: Okay cool, I'm good without it then

You can write an Extended JSON string from a list of dictionaries
by calling the ``bson.json_util.dumps()`` method. 
The following example outputs an Extended JSON string in the Relaxed format:

.. io-code-block::

   .. input::
      :language: python

      from bson import Code, Binary
      from bson.json_util import dumps

      doc = [
         {'foo': [1, 2]},
         {'bar': {'hello': 'world'}},
         {'code': Code('function x() { return 1; }', {})},
         {'bin': Binary(b'\x01\x02\x03\x04', 128)}
      ]

      ejson_str = dumps(doc)
      print(ejson_str)

   .. output::
      :language: none
      :visible: false

      '''[
         {"foo": [1, 2]},
         {"bar": {"hello": "world"}},
         {"code": {
            "$code": "function x() { return 1; }",
            "$scope": {}
         }},
         {"bin": {
            "$binary": {
               "base64": "AQIDBA==",
               "subType": "80"
         }}}
      ]'''

> By default, the ``dumps()`` method returns the Extended JSON string in the Relaxed

Reviewer 2: What is the Relaxed format? Would users be reasonably expected to know what that means without prior exposure?

Author: I tried to keep this page focused on just the PyMongo aspects of it to avoid redundancy with the server docs (linked on this page). That said, both you and the docs reviewer asked whether this page should include background info on the format, which makes me think we should 😃

-  ``LEGACY_JSON_OPTIONS``: Returns the Extended JSON string in Legacy format.
   We recommend using Relaxed or Canonical format instead.

The following example shows how to output Extended JSON in the Canonical format:

.. io-code-block::

   .. input::
      :language: python

      from bson import Code, Binary
      from bson.json_util import dumps, CANONICAL_JSON_OPTIONS

      doc = [
         {'foo': [1, 2]},
         {'bar': {'hello': 'world'}},
         {'code': Code('function x() { return 1; }', {})},
         {'bin': Binary(b'\x01\x02\x03\x04', 128)}
      ]

      ejson_str = dumps(doc, json_options=CANONICAL_JSON_OPTIONS)
      print(ejson_str)

   .. output::
      :language: none
      :visible: false

      '''[
         {"foo": [
            {"$numberInt": "1"},
            {"$numberInt": "2"}
         ]},
         {"bar": {"hello": "world"}},
         {"code": {
            "$code": "function x() { return 1; }",
            "$scope": {}
         }},
         {"bin": {
            "$binary": {
               "base64": "AQIDBA==",
               "subType": "80"
         }}}
      ]'''

More Information

> ----------------

Author: I wish our docs used 'More' because it seems shorter and simpler, but 🤷‍♂️
